File: Greedy/gas-station.py
from typing import List
# 64ms 60.07%
# 14.3MB 53.35%
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        delta = [gas[i] - cost[i] for i in range(len(gas))]
        if len(delta) == 0:
            return -1
        elif len(delta) == 1:
            if delta[0] >= 0:
                return 0
            else:
                return -1
        left = 0
        right = 0
        total = delta[left]
        while True:
            if total < 0:
                if left == 0:
                    left = len(delta) - 1
                else:
                    left -= 1
                total += delta[left]
            else:
                right = (right + 1) % (len(delta))
                total += delta[right]
            if left == right:
                break
        if total >= 0:
            return left
        else:
            return -1

a = Solution()
gas  = [1,2]
cost = [2,1]
print(a.canCompleteCircuit(gas, cost))

File: Greedy/gas-station.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        vector<int> remain_gas;
        remain_gas.push_back(gas[0]-cost[0]);
        for (int i=1; i<gas.size(); i++){
            remain_gas.push_back(remain_gas[i-1]+gas[i]-cost[i]);
        }
        if (remain_gas.back() < 0) return -1;

        auto min_value = min_element(remain_gas.begin(), remain_gas.end());
        int min_index = distance(remain_gas.begin(), min_value);
        return (min_index + 1) % gas.size();
    }
};

int main() {
    Solution sol;
    vector<int> gas = {5,8,2,8};
    vector<int> cost = {6,5,6,6};
    cout << sol.canCompleteCircuit(gas, cost) << endl;
    return 0;
}